fastplotlib/layouts/_engine.py

```diff
@@ -0,0 +1,384 @@
+            # ignore delta_x
+            new_extent[:2] = self._active_subplot.frame.extent[:2]
+
+        if (new_extent[2:] < 0).any():
+            # ignore delta_y
+            new_extent[2:] = self._active_subplot.frame.extent[2:]
+
+        # canvas extent
+        cx0, cy0, cw, ch = self._canvas_rect
+
+        # check if new x-range is beyond canvas x-max
+        if (new_extent[:2] > cx0 + cw).any():
+            new_extent[:2] = self._active_subplot.frame.extent[:2]
+
+        # check if new y-range is beyond canvas y-max
+        if (new_extent[2:] > cy0 + ch).any():
+            new_extent[2:] = self._active_subplot.frame.extent[2:]
+
+        return new_extent
+
+    def _action_start(self, subplot: Subplot, action: str, ev):
+        if self._inside_render_rect(subplot, pos=(ev.x, ev.y)):
+            return
+
+        if ev.button == 1:  # left mouse button
+            self._active_action = action
+            if action == "resize":
+                subplot.frame.resize_handle.material.color = (
+                    subplot.frame.resize_handle_color.action
+                )
+            elif action == "move":
+                subplot.frame.plane.material.color = subplot.frame.plane_color.action
+            else:
+                raise ValueError
+
+            self._active_subplot = subplot
+            self._last_pointer_pos[:] = ev.x, ev.y
+
+    def _action_iter(self, ev):
+        if self._active_action is None:
+            return
+
+        delta_x, delta_y = self._last_pointer_pos - (ev.x, ev.y)
+        new_extent = self._new_extent_from_delta((delta_x, delta_y))
+        self._active_subplot.frame.extent = new_extent
+        self._last_pointer_pos[:] = ev.x, ev.y
+
+    def _action_end(self, ev):
+        self._active_action = None
+        if self._active_subplot is not None:
+            self._active_subplot.frame.resize_handle.material.color = (
+                self._active_subplot.frame.resize_handle_color.idle
+            )
+            self._active_subplot.frame.plane.material.color = (
+                self._active_subplot.frame.plane_color.idle
+            )
+        self._active_subplot = None
+
+        self._last_pointer_pos[:] = np.nan
+
+    def set_rect(self, subplot: Subplot, rect: tuple | list | np.ndarray):
+        """
+        Set the rect of a Subplot
+
+        Parameters
+        ----------
+        subplot: Subplot
+            the subplot to set the rect of
+
+        rect: (x, y, w, h)
+            as absolute pixels or fractional
+
+        """
+
+        new_rect = RectManager(*rect, self._canvas_rect)
+        extent = new_rect.extent
+        # check for overlaps
+        for s in self._subplots:
+            if s is subplot:
+                continue
+
+            if s.frame.rect_manager.overlaps(extent):
+                raise ValueError(f"Given rect: {rect} overlaps with another subplot.")
+
+    def set_extent(self, subplot: Subplot, extent: tuple | list | np.ndarray):
```

Comment:
Interesting idea to have these two flavors to define a rectangular region! ❤️

Reply:
Yes! I find both useful in different cases.